cou/apps/base.py

```diff
@@ -205,9 +207,16 @@ def channel_codename(self) -> OpenStackRelease:
     def current_os_release(self) -> OpenStackRelease:
         """Current OpenStack Release of the application.
 
+        Applications that are colocated in a same machine will upgrade all the packages in the
+        machine during an upgrade. This means that when upgrading a application X, the application
+        Y colocated with it also upgrades its packages. To ensure to change the 'source' or
+        'openstack-origin' and run all the upgrade steps necessary, it's necessary to include the
+        OpenStack release set in the application configuration.
         :return: OpenStackRelease object
         :rtype: OpenStackRelease
         """
+        if self.os_origin:
+            return min(list(self.os_release_units.keys()) + [self.apt_source_codename])
         return min(self.os_release_units.keys())
```

Comment:
This seems fragile... this if seems to only exist to avoid an infinite loop with line 156. In other words the implementation of `apt_source_codename` leaks into `current_os_release` and vice versa. I suggest changing `apt_source_codename` to treat `os_origin` being empty or absent or set to `distro` in the same way, and dropping the if here. Also note https://github.com/juju/charm-helpers/blob/807f705d1d132bedb62b2eb743403e1d5867360f/charmhelpers/fetch/ubuntu.py#L654